src/ff/prime_field.rs

```diff
@@ -112,6 +116,60 @@ macro_rules! field_impl {
                 *self == other.as_u128()
             }
         }
+
+        #[cfg(test)]
+        paste! {
+            #[cfg(test)]
+            mod [< $field:lower _tests >] {
+                use super::*;
+                use proptest::proptest;
+
+                #[test]
+                fn zero() {
+                    let prime = u128::from($field::PRIME);
+                    assert_eq!($field::ZERO, $field::from(prime), "from takes a modulus",);
+                    assert_eq!($field::ZERO, $field::ZERO + $field::ZERO);
+                    assert_eq!($field::ZERO, $field::ZERO - $field::ZERO);
+                    assert_eq!($field::from(prime - 1), $field::ZERO - $field::ONE);
+                    assert_eq!($field::ZERO, $field::ZERO * $field::ONE);
+                }
+
+                #[test]
+                fn ser_not_enough_capacity() {
+                    let mut buf = [0; 0];
+                    assert!(matches!(
+                        $field::ONE.serialize(&mut buf),
+                        Err(e) if e.kind() == std::io::ErrorKind::WriteZero));
+                }
+
+                #[test]
+                fn de_buf_too_small() {
+                    let mut buf = [0; 0];
+                    assert!(matches!(
+                                    $field::deserialize(&mut buf),
+                                    Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof));
+                }
+
+                #[test]
+                fn can_write_into_buf_larger_than_required() {
+                    let mut buf = vec![0_u8; $field::SIZE_IN_BYTES as usize + 1];
+
+                    // panic will show the error while assert will just tell us that something went wrong
+                    $field::ONE.serialize(&mut buf).unwrap();
+                }
+
+                proptest! {
```

Comment:
How much work does this take when it passes?  It probably doesn't take much, but what assurance do we have?  An exhaustive search on a large field would be awkward.

Reply:
it is pretty fast as proptest doesn't really do an exhaustive search. It picks values randomly from the provided range and has configurable number of iterations (default is 100 or 1000). 

```bash
cargo test --lib ff::prime_field  -- -Zunstable-options --report-time

running 20 tests
test ff::prime_field::fp2::common_tests::zero ... ok <0.000s>
test ff::prime_field::fp31::common_tests::can_write_into_buf_larger_than_required ... ok <0.000s>
test ff::prime_field::fp2::specialized_tests::fp2 ... ok <0.000s>
test ff::prime_field::fp2::specialized_tests::fp2_binary_op ... ok <0.000s>
test ff::prime_field::fp2::common_tests::can_write_into_buf_larger_than_required ... ok <0.000s>
test ff::prime_field::fp2::common_tests::de_buf_too_small ... ok <0.000s>
test ff::prime_field::fp31::common_tests::de_buf_too_small ... ok <0.000s>
test ff::prime_field::fp31::common_tests::ser_not_enough_capacity ... ok <0.000s>
test ff::prime_field::fp2::common_tests::ser_not_enough_capacity ... ok <0.000s>
test ff::prime_field::fp31::specialized_tests::fp31 ... ok <0.000s>
test ff::prime_field::fp31::common_tests::zero ... ok <0.000s>
test ff::prime_field::fp32bit::common_tests::can_write_into_buf_larger_than_required ... ok <0.000s>
test ff::prime_field::fp32bit::common_tests::de_buf_too_small ... ok <0.000s>
test ff::prime_field::fp32bit::common_tests::ser_not_enough_capacity ... ok <0.000s>
test ff::prime_field::fp32bit::common_tests::zero ... ok <0.000s>
test ff::prime_field::fp32bit::specialized_tests::thirty_two_bit_additive_wrapping ... ok <0.000s>
test ff::prime_field::fp32bit::specialized_tests::thirty_two_bit_prime ... ok <0.000s>
test ff::prime_field::fp2::common_tests::serde ... ok <0.008s>
test ff::prime_field::fp31::common_tests::serde ... ok <0.008s>
test ff::prime_field::fp32bit::common_tests::serde ... ok <0.008s>
```